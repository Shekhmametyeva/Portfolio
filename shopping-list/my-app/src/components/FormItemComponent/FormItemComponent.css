.shopping__item__form {
    position: absolute;
    top: 0;
    left: 0;
    padding: 0;
}
.loading {
    width: 40px;
    height: 40px;
    background: url(../../assets/icons/load.gif) 0/cover;
}