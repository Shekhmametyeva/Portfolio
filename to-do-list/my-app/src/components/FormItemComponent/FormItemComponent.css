.todo__item__form {
    padding: 8px 20px;
    position: absolute;
    top: 0;
    left: 0;
    
}
.todo__item__form input {
    font-size: 1.2rem;
}

