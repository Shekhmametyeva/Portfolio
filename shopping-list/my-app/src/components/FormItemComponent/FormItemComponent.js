import './FormItemComponent.css';
import React from 'react';
import ButtonSvgComponents from '../buttonSvgComponent/ButtonSvgComponent';

class FormItemComponent extends React.Component {
    constructor(props) {
        super(props);
        this.editData = this.editData.bind(this)
    }
    editData() {
        
    }
    handleSubmit(event){
        event.preventDefault();
        this.props.callback.editData(this.props.element.id, this.props.value);
        this.props.callback.openForm()
    }
    render() {
        return(
            <form className="shopping__form shopping__item__form" onSubmit={(event) => this.handleSubmit(event)}>
                <input 
                autoFocus 
                type='text' 
                value={this.props.value} 
                placeholder={this.props.element.title} 
                className='shopping__input' 
                onChange={(event) => {this.props.callback.changeStateitemValue(event.target.value)}}/>
                <ButtonSvgComponents type='submit' name='add' class='shopping__icon' />
                <ButtonSvgComponents name='add' class='close shopping__icon' callback={this.props.callback.openInput}/>
            </form>
        )
    }
}
export default FormItemComponent
