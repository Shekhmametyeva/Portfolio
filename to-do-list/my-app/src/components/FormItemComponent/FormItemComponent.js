import './FormItemComponent.css';
import ButtonSvgComponent from '../ButtonSvgComponent/ButtonSvgComponent';
import React, { useState } from 'react';

function checkValidity (props, value, setValue, setValid) {
    console.log(props.dataFull)
    if(!value.trim()) {
        setValid('поле не заполнено')
        setTimeout(() => {
            setValue()
            setValid(false)
        }, 1500);
        return null
    }
    if(props.element === value.trim()) {
        props.closeInput()
        return null
    }
    if(props.dataFull.find(el => el.value === value.trim())) {
        props.highlight(value.trim())
        setTimeout(() => {
            setValue()
        }, 1500);
        return null
    }
    setValue()
    return value.trim()
}

function FormItemComponent (props) {
    const [value, setValue] = useState('');
    const [valid, setValid] = useState(false);
    const prompt = valid ? <p className='valid'>{valid}</p> : null
    return (
        <form className='todo__form todo__item__form' 
            onSubmit={(event) => {
                event.preventDefault();
                const newValue = checkValidity(props, value, () => setValue(''), (text)=> setValid(text));
                if (newValue) {
                    props.updateStateData(props.index, newValue)
                }
                
            }}> 
            <div className='input__container'>
                <input
                    autoFocus 
                    readOnly={valid}
                    type='text' 
                    value={value} 
                    className='todo__input' 
                    onChange={(event) => {
                        setValue(event.target.value)
                    }}
                />
                <ButtonSvgComponent class='todo__button' name='add' type='submit' disabled={valid}/>
            </div>
            {prompt}
        </form>
    )

}
export default FormItemComponent