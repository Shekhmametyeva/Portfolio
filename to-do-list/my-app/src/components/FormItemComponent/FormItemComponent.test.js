import {render, screen, fireEvent} from '@testing-library/react';
import FormItemComponent from './FormItemComponent';

describe('тестируем FormItemComponent', () => {
    test('render FormItemComponent', () => {
        render(<FormItemComponent />);
        const formItem = screen.getByTestId('form-item');
        const input = screen.getByRole('textbox')
        expect(input).toBeInTheDocument();
        expect(formItem).toMatchSnapshot()
    })
    test('при пустом вводе появляется "поле не заполнено"', () => {
        render(<FormItemComponent />);
        const input = screen.getByRole('textbox');
        const btn = screen.getByTestId('btn-edit-elem');
        fireEvent.input(input, {target: {value: ' '}});
        fireEvent.click(btn);
        expect(input).toBeInTheDocument();
        const errorText = screen.getByText(/поле не заполнено/i) 
        expect(errorText).toBeInTheDocument()
    })
})