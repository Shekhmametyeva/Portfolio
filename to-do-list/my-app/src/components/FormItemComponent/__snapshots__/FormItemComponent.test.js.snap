// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`тестируем FormItemComponent render FormItemComponent 1`] = `
<form
  class="todo__form todo__item__form"
  data-testid="form-item"
>
  <div
    class="input__container"
  >
    <input
      class="todo__input"
      type="text"
      value=""
    />
    <button
      class="todo__button button"
      data-testid="btn-edit-elem"
      type="submit"
    >
      <svg
        class="icon"
      >
        <use
          xlink:href="sprite.svg#add"
        />
      </svg>
    </button>
    <button
      class="todo__button close button"
      type="button"
    >
      <svg
        class="icon"
      >
        <use
          xlink:href="sprite.svg#add"
        />
      </svg>
    </button>
  </div>
</form>
`;
